src/Accelerators/NNPA/Conversion/ZHighToZLow/ZHighToZLow.cpp

```diff
@@ -1580,8 +1585,135 @@ struct ZHighToZLowStickifiedConstantOfShapeOpLowering
   }
 };
 
+//===----------------------------------------------------------------------===//
+// A template to lower ZHigh DLF16ToF32 and F32ToDLF16.
+//===----------------------------------------------------------------------===//
+
+template <typename CONVERT_OP>
+struct ZHighToZLowDataConversionLowering
+    : public OpConversionPattern<CONVERT_OP> {
+  using OpAdaptor = typename CONVERT_OP::Adaptor;
+  bool fromF32 = false;
+  bool enableParallel = false;
+
+  ZHighToZLowDataConversionLowering(TypeConverter &typeConverter,
+      MLIRContext *ctx, bool fromF32, bool enableParallel)
+      : OpConversionPattern<CONVERT_OP>(typeConverter, ctx), fromF32(fromF32),
+        enableParallel(enableParallel) {}
+
+  LogicalResult matchAndRewrite(CONVERT_OP convertOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    Location loc = convertOp.getLoc();
+    MDBuilder create(rewriter, loc);
+
+    Operation *op = convertOp.getOperation();
+    ValueRange operands = adaptor.getOperands();
+    Value X = operands[0];
+    int64_t rank = getRank(X.getType());
+
+    // SIMD info.
+    // Fixed VL for the conversion instruction: 8 elements per instruction call.
+    int64_t VL = 8;
```

Comment:
Note, we can get better loop unrolling by selecting VL as multiple of the machine VL; with normal operations, an `add` of vector length 16 becomes 2 simd instructions. So for example, you could unroll by a cache length, so that you get better register usage by having multiple instruction in flight at the same time and (for parallel) no false sharing. What I have done is estimate the register pressure... so if it's simple like here, you can definitely have 4-8 instructions in flight. If has a lot of intermediary results, then its better not to unroll too much as we don't want spilling. In any case, the backend may also unroll, if you do it here, you are sure it happens :-)

Reply:
Yes, you are right. I have another round for optimization, e.g. optimizing LayoutTransform, so I will do unrolling then with a new PR.